### ISSUE TITLE
18.15 add anti-automation solution.

### ISSUE THREAD

**Issue by tghosth** [OWASP/ASVS#337]
18.15 currently says:
> If no access control is present, verify that the REST service isn't susceptible to excessive calls, which could lead to increased overhead or CPU/bandwidth bills. A potential solution could be API keys, **or returning error code 429 "Too Many Requests" if requests are excessive.**

That last part doesn't feel like a solution but rather just what error message to return. Could I suggest:
> If no access control is present, verify that the REST service isn't susceptible to excessive calls, which could lead to increased overhead or CPU/bandwidth bills. A potential solution could be API keys, **or implementing anti-automation techniques such as rate limiting.**

It is not perfect but at least it points them in the right direction.

**Comment by mastacheata** [comment]
Isn't rate limiting just one part of anti-automation?  
You could perform very basic rate limiting (i.e. limit # of requests/IP) or some sort of adaptive rate limiting based on other risk factors (i.e.: Block users completely if the IP is known to belong to a datacenter instead of a business/consumer ISP).
Another thing worth thinking about is load balancing and/or high availability.

**Comment by tghosth** [comment]
So this specific control discussed excessive calls, hence specifically mentioning rate limiting.

Load balancing and/or high availability is interesting but I wonder if that should be in V19. What do you think @mastacheata ? Can you think of specific controls to add there?

**Comment by mastacheata** [comment]
> Load balancing and/or high availability is interesting but I wonder if that should be in V19. What do you think @mastacheata ? Can you think of specific controls to add there?

The term "Load balancing" would fit right with the excessive requests stuff, but it's probably not security relevant.
High availability on the other hand would definitely fit in V19, despite being implemented in very similar if not the same way.

I'll open another issue with an idea for a rule for high availability.

**Comment by tghosth** [comment]
> Understood. Maybe reword it like this: 
> implementing anti-automation techniques like rate limiting.

Done, see original comment

**Comment by tghosth** [comment]
Committed
